Google.Api.Generator.Tests/ResourcePatternTest.cs

```diff
@@ -14,28 +14,91 @@
         {
             var pat = new ResourcePattern(pattern);
             Assert.Equal(paramNames, pat.ParameterNames);
             Assert.Equal(pathTemplateString, pat.PathTemplateString);
+        }
+
+        [Theory]
+        [InlineData("as/{a}", new[] { "A" }, "as/A")]
+        [InlineData("as/{a=*}", new[] { "A" }, "as/A")]
+        [InlineData("as/{a=**}", new[] { "A/AA/AAA" }, "as/A/AA/AAA")]
+        [InlineData("as/{a=*}/bs/{b=*}/cs/{c=**}", new[] { "A", "B", "C/CC/CCC" }, "as/A/bs/B/cs/C/CC/CCC")]
+        [InlineData("as/{a_1}~{a_2}.{a_3}", new[] { "A1", "A2", "A3" }, "as/A1~A2.A3")]
+        [InlineData("as/{a_1=*}~{a_2=*}.{a_3=*}", new[] { "A1", "A2", "A3" }, "as/A1~A2.A3")]
+        [InlineData("as/{a=*}/bs/{b_1}-{b_2}/cs/{c_1=*}_{c_2}/{d=**}", new[] { "A", "B1", "B2", "C1", "C2", "D" }, "as/A/bs/B1-B2/cs/C1_C2/D")]
+        public void ValidExpansion(string pattern, string[] expandArgs, string expanded)
+        {
+            var pat = new ResourcePattern(pattern);
             Assert.Equal(expanded, pat.Expand(expandArgs));
         }
 
+        [Theory]
+        [InlineData("as/bs/cs", "as/bs/cs")]
+        [InlineData("*", "[^/]+")]
+        [InlineData("{a=*}", "([^/]+)")]
+        [InlineData("**", ".*")]
+        [InlineData("{a=**}", "(.*)")]
+        [InlineData("as/*", "as/[^/]+")]
+        [InlineData("as/*/bs/**", "as/[^/]+/bs(?:/.*)?")]
+        [InlineData("as/{a}", "as/([^/]+)")]
+        [InlineData("as/{a=*}", "as/([^/]+)")]
+        [InlineData("as/{a=**}", "as(/.*)?")]
+        [InlineData("as/{a=*}/bs/{b=**}", "as/([^/]+)/bs(/.*)?")]
+        [InlineData("{a=as/*}/{b=bs/**}", "(as/[^/]+)/(bs(?:/.*)?)")]
+        public void ValidRegexes(string pattern, string regexStr)
+        {
+            var pat = new ResourcePattern(pattern);
+            Assert.Equal(regexStr, pat.RegexString);
+        }
+
+        /// <summary>
+        /// This patterns in this test are used in the tests in GAX (RoutingExtractorTest).
+        /// It is therefore intentionally contains similar test cases.
+        /// </summary>
+        [Theory]
+        [InlineData("{table_name=projects/*/instances/*/**}", "^(projects/[^/]+/instances/[^/]+(?:/.*)?)$")]
+        [InlineData("{table_name=regions/*/zones/*/**}", "^(regions/[^/]+/zones/[^/]+(?:/.*)?)$")]
+        [InlineData("{routing_id=projects/*}/**", "^(projects/[^/]+)(?:/.*)?$")]
+        [InlineData("{routing_id=projects/*/instances/*}/**", "^(projects/[^/]+/instances/[^/]+)(?:/.*)?$")]
+        [InlineData("{project_id=projects/*}/instances/*/**", "^(projects/[^/]+)/instances/[^/]+(?:/.*)?$")]
+        [InlineData("projects/*/{instance_id=instances/*}/**", "^projects/[^/]+/(instances/[^/]+)(?:/.*)?$")]
+        [InlineData("{project_id=projects/*}/**", "^(projects/[^/]+)(?:/.*)?$")]
+        [InlineData("subs/{sub.sub_name}", "^subs/([^/]+)/?$")]
+        [InlineData("{legacy.routing_id=**}", "^(.*)$")]
```

Comment:
Updated

Reply:
I'm not seeing this one being updated. I saw the other comment about only some cases wanting `+` instead of `*` - I'll take a more careful look before merging.